db/range_tombstone_fragmenter.cc

```diff
@@ -0,0 +1,223 @@
+//  Copyright (c) 2018-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "db/range_tombstone_fragmenter.h"
+
+#include <set>
+#include <algorithm>
+#include <functional>
+
+#include <stdio.h>
+#include <inttypes.h>
+
+#include "util/kv_map.h"
+
+namespace rocksdb {
+
+namespace {
+
+struct ParsedInternalKeyComparator {
+  ParsedInternalKeyComparator(const InternalKeyComparator* c) : cmp(c) {}
+
+  bool operator()(const ParsedInternalKey& a,
+                  const ParsedInternalKey& b) const {
+    return cmp->Compare(a, b) < 0;
+  }
+
+  const InternalKeyComparator* cmp;
+};
+
+}  // anonymous namespace
+
+FragmentedRangeTombstoneIterator::FragmentedRangeTombstoneIterator(
+    std::unique_ptr<InternalIterator> unfragmented_tombstones,
+    const InternalKeyComparator& icmp, SequenceNumber snapshot)
+    : tombstone_cmp_(icmp.user_comparator()),
+      icmp_(&icmp),
+      ucmp_(icmp.user_comparator()) {
+  if (unfragmented_tombstones == nullptr) {
+    pos_ = tombstones_.end();
+    return;
+  }
+  Slice cur_start_key(nullptr, 0);
+  auto cmp = ParsedInternalKeyComparator(&icmp);
+
+  // Stores the end keys and sequence numbers of range tombstones with a start
+  // key of cur_start_key. Provides an ordering by end key for use in
+  // flush_current_tombstones.
+  std::set<ParsedInternalKey, ParsedInternalKeyComparator> cur_end_keys(cmp);
+
+  // Given the next start key in unfragmented_tombstones,
+  // flush_current_tombstones writes every tombstone fragment that starts and
+  // ends with a key before next_start_key.
+  auto flush_current_tombstones = [&](const Slice& next_start_key) {
+    auto it = cur_end_keys.begin();
+    bool reached_next_start_key = false;
+    for (; it != cur_end_keys.end() && !reached_next_start_key; ++it) {
+      Slice cur_end_key = it->user_key;
+      if (icmp.user_comparator()->Compare(cur_start_key, cur_end_key) ==
+          0) {
+        // Empty tombstone.
+        continue;
+      }
+      if (icmp.user_comparator()->Compare(next_start_key, cur_end_key) <= 0) {
+        reached_next_start_key = true;
+        cur_end_keys.erase(cur_end_keys.begin(), it);
```

Comment:
Does this invalidate the iterator? I think you need to use the return value of this call.